### PR TITLE
storage: add bare-bones UNITTEST source

### PR DESCRIPTION
**Pull request by guswynn**
This pr adds a new source `UNITTEST`, that lets the user describe a script of values, offset, keys, etc. that the source emits in turn.

I chose to go with the simplest possible route: a source that emits values and possibly keys, of just bytes, just like kafka. This means that in theory, every format and envelope is supported (in practice, some things around avro confluent schemas are broken).

This is the bare-bones implementation, and supports only keys, offsets, values, and terminating and non-terminating sources. Additionally, I went with the most basic implementation: a json string in the source, un-validated by anything but the source-reader. In the future, we likely want to validate this json (but for the proto transport, the json is still probably the best way to transfer the script). I chose this because this is the the most perma-unstable source, and I want people refactoring to have as little to do to fix it up as possible.

In the future, features I'd like to see:
- controlled delays before and after value emission
- transactions (i.e. `InProgress` messages)
- `INCLUDE OFFSET` working
- control over how _batches in the source pipeline are formed_ (this requires an escape hatch through the `SourceReader` trait)
- (maybe) assertions, including assertions about committed offsets.
- a cargo-test compatible driver for source rendering (which this source will drop perfectly into)
  - ^ is what I want to work on next, but its non-trivia
  - 
### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

### CODE REVIEW

File: test/testdrive/unittest_source.td [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+> CREATE CONNECTION c_conn
+  FOR CONFLUENT SCHEMA REGISTRY URL '${testdrive.schema-registry-url}'
+
+> CREATE SOURCE unit
+  FROM UNITTEST
+  '[
+    {"key": "fish", "val": "value", "offset": 0},
+    {"key": "fish2", "val": "hmm", "offset": 1},
+    {"key": "fish", "val": "value2", "offset": 2, "pend_forever": true}
+  ]'
```
Comment (benesch) [inline test/testdrive/unittest_source.td:22]: Can you add documentation somewhere about the syntax + semantics of the test script? Also, I recommend going with a more command-based approach:

```rust
#[serde(tag = "command")]
#[serde(rename_all = "lowercase")]
enum ScriptCommand {
    Emit { key: String, value: String, offset: ... },
    Block,
}

{ "command": "emit",  "key": "fish", "value": "value", "offset": 0}
{ "command": "block" }
```

**Review by benesch — COMMENTED** [review]
Turns out I do have some thoughts after all!

First, can we call this the `TEST SCRIPT` source rather than the `UNITTEST` source? I think it does a better job capturing what the source is—it might not just be unittests! Also, separate keywords for separate words per our new SQL design principles document: https://www.notion.so/materialize/SQL-design-principles-2e2069797f7c46b8b4c7dd10043afdcd#9d4a155e751f4cbeab7988bfb338b3f6

**Review by philip-stoev — APPROVED** [review]
So basically this is like creating a kafka source and doing a `kafka-ingest`, but with more control over timestamps and such?

Did you envision this for unittest-level testing or for something larger? We could do several things to increase the scope of the testing:
- I can pseudo-randomly generate sequences of commands and feed them through this (with limited opportunity to validate the data returned from the source though), if you think the product should be successful able to process any sequence without panick-ing.
- if you also add some sort of `repeat=N` or `size` parameter, this could be used to cause very large volumes of data to be ingested for the purpose of stress testing. Right now we create large volumes with either testdrive's `kafka-ingest repeat=XXXX` or a separate kafka message generator, and both of those methods have overheads that reduce the stressfullness of the entire test
- or the source can have an option `emit forever` that will just continue to emit the same commands over and over again until `DROP`

**Comment by guswynn** [comment]
@philip-stoev 

I want to use this for actual unit-testing of the source pipeline (https://github.com/MaterializeInc/materialize/pull/15430). Exposing to sql is not strictly required but seemed nice. Also, in addition to control over offset, I eventually want give this source a bit more control into the source pipeline to let me construct batches of a specific shape!

The goal is to be able to write easy to iterate, non-timing-dependent regression tests for things like the capability bug!

**Comment by guswynn** [comment]
@philip-stoev your ideas for additional testing are all great! We have technically a very small number of constraints for messages coming out of sources (offsets must stay = or go up, `SourceMessageType::Finalized` can't lie), so if we follow those, anything else should work!

> if you also add some sort of repeat=N or size parameter,

I specifically chose the json-based representation to make it as EASY as possible to add new options/features to this source, for your sake and my team's sake

Hopefully today i can get to nikhil's review and get this merged

**Pull request merged**